# Unidad 5: Archivos

<img src="img\broken2.jpg" alt="Python" width="1000"/>

Un archivo puede que sea diferente a lo que esta uno acostumbrado a manejar en Windows. En Python cuando se hace referencia a un archivo, se está refiriendo a un documento de texto (txt) o a un documento binario. El primero corresponde a una secuencia de lineas formadas por caracteres. En cambio los archivos binarios debido a su naturaleza, es necesario usar alguna aplicación para poder leerlos o interpretarlos.

En esencia, los archivos son un conjunto de bytes que se utilizan para almacenar cierto tipo de datos. En la actualidad, los archivos en la mayoría de los sistemas operativos están compuestos de 3 partes:

1. **Encabezado**: Metadatos acerca del contenido del archivo (nombre, tamaño, tipo, etc.)

2. **Datos**: El contenido del archivo.

3. **Fin del archivo (EOF)**: Un caracter especial que indica el final del documento. 

Dependiendo el tipo de archivo que se vaya a utilizar, es el tipo de extensión que lo representará. En este curso se manejarán archivos con extensión txt para textos y csv para bases de datos.

Al tratar de acceder a un archivo, es importante conocer la localización del mismo. Dicha localización estará compuesta de 3 secciones:

1. La ruta
2. El nombre del archivo
3. La extensión

Es posible detrminar si un archivo es de texto o si es binario basandonos únicamente en la extensión del mismo. A continuación se presentan ejemplos de las extensiones más comunes:

**Archivos Binarios:**

<table>
  <tr>
    <td>Imagenes</td>
    <td>jpg, png, gif, bmp, tiff, psd</td>
  </tr>
  <tr>
    <td>Videos</td>
    <td>mp4, mvk, avi, mov, mpg, vob</td>
  </tr>
  <tr>
    <td>Audio</td>
    <td>mp3, aac,wav, flac, ogg, mka, wma</td>
  </tr>
  <tr>
    <td>Documentos</td>
    <td>pdf, doc, xls, ppt, docx</td>
  </tr>
  <tr>
    <td>Archivo</td>
    <td>zip, rar, iso</td>
  </tr>
  <tr>
    <td>Ejecutable</td>
    <td>exe, dll</td>
  </tr>
</table>

**Archivos de texto:**

<table>
  <tr>
    <td>Web</td>
    <td>html, xml, css, svg, json</td>
  </tr>
  <tr>
    <td>Código fuente</td>
    <td>c, cpp, h, cs, js, py, java, rb, pl, php</td>
  </tr>
  <tr>
    <td>Documentos</td>
    <td>txt, tex, markdown, asciidoc</td>
  </tr>
  <tr>
    <td>Configuración</td>
    <td>ini, cfg, reg</td>
  </tr>
  <tr>
    <td>Tabular</td>
    <td>csv, tsv</td>
  </tr>
</table>

## 5.1 Procesamiento de archivos

El procesamiento de archivos está constituido por 3 sencillos pasos: **Abrir el archivo, Manipular el contenido, Cerrar el archivo.**

### Abrir el archivo

Abrir un archivo en Python se hace mediante el método **open()**. Para poder abrir un archivo es necesario conocer su ubicación y nombre seguido. Opcionalmente, por el modo o tipo de operación a realizar, se selecciona alguna codificación que nos servirá para manipular el archivo. Si no se indica el tipo de operación el archivo se abrirá en modo de lectura y si se omite la codificación se utilizará la codificación actual del sistema. Si se intenta abrir para un archivo inexistente, Python creará uno nuevo. Los modos de apertura en Python son los siguientes:

<table>
  <tr>
    <th>Modificador</th>
    <th>Significado</th>
  </tr>
  <tr>
    <td> r </td>
    <td>Abre un documento para su lectura (esta es la opción por defecto)</td>
  </tr>
  <tr>
    <td> w </td>
    <td>Abre un documento para su escritura (en caso de no existir creará el archivo)</td>
  </tr>
  <tr>
    <td> a </td>
    <td>Abre un archivo para agregar texto al final del documento sin sobreescribirlo</td>
  </tr>
  <tr>
    <td> t </td>
    <td>Abre un documento en modod de texto (esta es la opción por defecto)</td>
  </tr>
  <tr>
    <td> b </td>
    <td>Abre un documento en modo binario</td>
  </tr>
  <tr>
    <td> + </td>
    <td>Abre un documento para ser actualizado (lectura y escritura)</td>
  </tr>
</table>

### Manipular el archivo

Manipular archivos en Python es muy sencillo. No hace falta importar ninguna librería. Los métodos para modificar los archivos son **read()** y **write()**. El primero nos permite leer la información almacenada en el documento mientras que el segundo nos permíte agregar información. Cabe mencionar que mediante esta técnica, no es posible modificar texto ya existente, solamente es posible agregar nuevas lineas de texto. En caso de querer modificar el texto, sería necesario reescribir todo el archivo en cuestión. 

Es posible desplazarse en el texto. Para ello se utilizan los métodos **tell()** y **seek()**. El primero indicará la posición del cursor en bytes. El segundo desplazara al cursor a la posición indicada en bytes. Dicha posición está indexada.

### Cerrar el archivo

Python, por tratarse de un lenguaje de programación, utiliza recursos de memoria siempre que una función o un método es ejecutado. En este sentido, abrir un archivo y manipularlo utiliza recursos. Estos recursos son liberados solo hasta que el documento se cierra. Para cerrar un archivo se utiliza el método **close()**, y al momento de cerrar el archivo, todos los cambios son guardados.

### try - finally / with - as

Al estar manipulando un archivo en Python, es probable que una excepción pueda ocurrir.  Cuando una de estas pasa, el código saldrá abruptamente provocando que el archivo se cierre sin previo aviso. Una manera de controlar esta situación es mediante el manejo de excepciones. El bloque **try - finally** es una manera mas segura de abrir un archivo:

In [ ]:
try:
    f = open("test.txt","w+")
    # Operaciones a realizar en el archivo
finally:
    f.close()

De este modo, garantizamos que el archivo se cierre de forma adecuada, incluso si una excepción llegara a ocurrir. Para poder lograr el manejo de excepción de arriba, Python utiliza el bloque **with - as**

In [ ]:
with open("test.txt", "a") as f:
    # Operaciones a realizar en el archivo

### Ejemplo: Abrir, manipular y cerrar:

In [ ]:
with open("test.txt",'w+',encoding = 'utf-8') as f:
    f.write("Mi primer archivo\n")
    f.write("Este archivo\n\n")
    f.write("contiene 4 lineas\n\n")
    f.write("Adios\n")

El ejemplo anterior, abre el archivo test, en caso de no existir lo crea, y en caso de existir sobreescribira el contenido con las lineas escritas dentro del bloque.

Repitiendo el ejemplo anterior, ahora usaremos el método **tell()** para determinar la posición final del cursor:

In [ ]:
with open("test.txt",'w+',encoding = 'utf-8') as f:
    f.write("Mi primer archivo\n")
    f.write("Este archivo\n\n")
    f.write("contiene 4 lineas\n\n")
    f.write("Adios\n")
    a=f.tell() # Como el documento cierra al término del bloque, almacenamos el valor de tell
               # en una variable, para posteriormente imprimirla y saber la última posición 
               # del cursor.
a

El número 63, representa el número de bytes del archivo. Cada caracter representa un byte, y cada salto de linea representa 2 bytes.

In [ ]:
with open("test.txt",'w+',encoding = 'utf-8') as f:
    f.write("\nMi primer archivo\n")
    f.write("Este archivo\n\n")
    f.write("contiene 4 lineas\n\n")
    f.write("Adios\n")
    f.seek(0)
    f.write("Hola\n")

## 5.2 Archivos de texto

En Python no hace falta importar alguna libreria para el manejo de archivos de texto. Esto se puede hacer directamente mediante los métodos ya incorporados en el lenguaje. El primero que ocuparemos es **open()**

-------------------------------------------------------------------------------------------------------------------------

                               open("Nombre del archivo.txt", "Permisos asignados")

-------------------------------------------------------------------------------------------------------------------------

El primer argumento representa el archivo que estamos tratando de abrir, mientras que el segundo representa el tipo de permiso u operación que queremos realizar en el archivo. Usando una letra **"w"** indica que queremos realizar una escritura en el documento. **"w+"** implica que además de la escritura, le pedimos a Python que genere el documento en caso de no existir. Vamos a ver como funciona:

In [ ]:
f= open("Texto_prueba.txt","w+")

Con esta instrucción, hemos creado un archivo con extensión txt en la misma carpeta en la que se encuentra almacenado el presente cuaderno Jupyter. En caso de querer utilizar algún otro permiso, los disponibles son: **"r"** para lectura y **"a"** para agregar.

In [ ]:
for i in range(10):
     f.write("Esta es la linea " + str(i+1) + "\r\n")

El ciclo anterior generó 10 lineas en nuestro archivo txt, cada una de esas lineas cuenta con su número correspondiente del ciclo. **\n** es para dar un salto de linea al termino del enunciado, y **\r** es para crear un espacio entre lineas. Por último, para que los cambios se efectuen, es necesario cerrar el documento en Python. Esto se consigue con el método **close()**:

In [ ]:
f.close()

Es posible agregar información al archivo, para ello utilizaremos otro modificador de permiso:

In [ ]:
f= open("Texto_prueba.txt","a")

In [ ]:
for i in range(2):
     f.write("Esta linea fue agregada después " + str(i+1) + "\r\n")

In [ ]:
f.close()

Finalmente, otra de las opciones para la manipulación de textos es la lectura. Para ello recurrimos al último modificador de permiso:

In [ ]:
f= open("Texto_prueba.txt","r")

Verificamos el status del documento mediante el atributo **mode()**:

In [ ]:
f.mode

Ya que nuestro documento efectivamente está en modo lectura, procedemos a hacer la lectura del mismo. Para ello usaremos el método **read()**. Por último almacenaremos en contenido en una variable para poder llamarla en cualquier momento:

In [ ]:
contenido=f.read()

In [ ]:
print(contenido)

Si revisamos la variable contenido, veremos que el formato que tiene es diferente al que se nos presenta en Python:

In [ ]:
contenido

Al tener el texto del archivo almacenado dentro de una variable, nos permite hacer uso de diferentes métodos para poderlo manipular. Un ejemplo de esto es el uso del método **split()**, el cual nos permite crear una lista de elementos al partir el texto en secciones de acuerdo al separador que le indiquemos:

In [ ]:
# El ciclo separá el contenido del texto cada que encuentre \n en él:

for i in contenido:
    lista_lineas = contenido.split("\n")
print (lista_lineas)

In [ ]:
len(lista_lineas)

La lista resultante tambien trajo los espacios generados por \r. Lo que podemos hacer ahora es filtrarlos de la lista resultante mediante el uso del método **remove()**:

In [ ]:
for i in lista_lineas:
    if i=='':
        lista_lineas.remove(i)

print (lista_lineas)

In [ ]:
lista_lineas.remove('')

In [ ]:
lista_lineas

In [ ]:
len(lista_lineas)

Podemos usar nuevamente el método **split()** pero ahora para crear una lista de las palabras contenidas en el texto:

In [ ]:
for i in contenido:
    lista_palabras = contenido.split(" ")
print (lista_palabras)

In [ ]:
len(lista_palabras)

### Actividad

Crea un documento txt, cuyo nombre sea Pangramas. Posteriormente, escribe cada una de las siguientes frases en líneas diferentes:

* Whisky bueno: ¡excitad mi frágil pequeña vejez!
* Quiere la boca exhausta vid, kiwi, piña y fugaz jamón
* Jovencillo emponzoñado de whisky: ¡qué figurota exhibe! (para la fuentes tipográficas bajo las distribuciones GNU/Linux)
* El viejo Señor Gómez pedía queso, kiwi y habas, pero le ha tocado un saxofón
* Mi hijo degustó en el festival de bayas una extraña pizza de kiwi con queso
* José compró una vieja zampoña en Perú. Excusándose, Sofía tiró su whisky al desagüe de la banqueta
* El cadáver de Wamba, rey godo de España, fue exhumado y trasladado en una caja de zinc que pesó un kilo
* El pingüino Wenceslao hizo kilómetros bajo exhaustiva lluvia y frío, añoraba a su querido cachorro
* El veloz murciélago hindú comía feliz cardillo y kiwi. La cigüeña tocaba el saxofón detrás del palenque de paja
* The quick brown fox jumps over the lazy dog

Posteriormente cierra el documento. 
Vuelve a abrir el archivo mediante el uso del bloque **with**, y agrega las siguientes 3 líneas al final del documento:

    (Salto de línea)

    (Salto de línea)

    "Un pangrama es un texto que usa todas las letras posibles del alfabeto de un idioma."


Despúes, indicar el número de bytes del documento mediante el uso del método **tell()**

Nuevamente abra el archivo en modo de lectura, y cuente el número de caracteres por renglón en el archivo. El resultado de cada renglon será imprimeso en python así como la suma total.

Obtenga la diferencia entre el número de bytes del archivo, y el número de caracteres obtenidos por renglon. **¿Porqué son diferentes?**

<img src="img/spyder.png" alt="Spyder" width="300"/>

In [5]:
actividad=f= open("Pangramas.txt","w+")

In [6]:
actividad.write("Whisky bueno: ¡excitad mi frágil pequeña vejez!\n")
actividad.write("Quiere la boca exhausta vid, kiwi, piña y fugaz jamón\n")
actividad.write("Jovencillo emponzoñado de whisky: ¡qué figurota exhibe! (para la fuentes tipográficas bajo las distribuciones GNU/Linux)\n")
actividad.write("El viejo Señor Gómez pedía queso, kiwi y habas, pero le ha tocado un saxofón\n")
actividad.write("Mi hijo degustó en el festival de bayas una extraña pizza de kiwi con queso\n")
actividad.write("José compró una vieja zampoña en Perú. Excusándose, Sofía tiró su whisky al desagüe de la banqueta\n")
actividad.write("El cadáver de Wamba, rey godo de España, fue exhumado y trasladado en una caja de zinc que pesó un kilo\n")
actividad.write("El pingüino Wenceslao hizo kilómetros bajo exhaustiva lluvia y frío, añoraba a su querido cachorro\n")
actividad.write("El veloz murciélago hindú comía feliz cardillo y kiwi. La cigüeña tocaba el saxofón detrás del palenque de paja\n")
actividad.write("The quick brown fox jumps over the lazy dog\n")

44

In [7]:
actividad.close()

In [19]:
with open("Pangramas.txt",'a') as actividad:
    actividad.write("\n\n'Un pangrama es un texto que usa todas las letras posibles del alfabeto de un idioma.'")
    a=actividad.tell()

In [20]:
a

932

In [22]:
actividad= open("Pangramas.txt","r")
contenido=actividad.read()
contenido

"Whisky bueno: ¡excitad mi frágil pequeña vejez!\nQuiere la boca exhausta vid, kiwi, piña y fugaz jamón\nJovencillo emponzoñado de whisky: ¡qué figurota exhibe! (para la fuentes tipográficas bajo las distribuciones GNU/Linux)\nEl viejo Señor Gómez pedía queso, kiwi y habas, pero le ha tocado un saxofón\nMi hijo degustó en el festival de bayas una extraña pizza de kiwi con queso\nJosé compró una vieja zampoña en Perú. Excusándose, Sofía tiró su whisky al desagüe de la banqueta\nEl cadáver de Wamba, rey godo de España, fue exhumado y trasladado en una caja de zinc que pesó un kilo\nEl pingüino Wenceslao hizo kilómetros bajo exhaustiva lluvia y frío, añoraba a su querido cachorro\nEl veloz murciélago hindú comía feliz cardillo y kiwi. La cigüeña tocaba el saxofón detrás del palenque de paja\nThe quick brown fox jumps over the lazy dog\n\n'Un pangrama es un texto que usa todas las letras posibles del alfabeto de un idioma.'"

In [23]:
for i in contenido:
    lista_lineas_pan = contenido.split("\n")
print (lista_lineas_pan)

['Whisky bueno: ¡excitad mi frágil pequeña vejez!', 'Quiere la boca exhausta vid, kiwi, piña y fugaz jamón', 'Jovencillo emponzoñado de whisky: ¡qué figurota exhibe! (para la fuentes tipográficas bajo las distribuciones GNU/Linux)', 'El viejo Señor Gómez pedía queso, kiwi y habas, pero le ha tocado un saxofón', 'Mi hijo degustó en el festival de bayas una extraña pizza de kiwi con queso', 'José compró una vieja zampoña en Perú. Excusándose, Sofía tiró su whisky al desagüe de la banqueta', 'El cadáver de Wamba, rey godo de España, fue exhumado y trasladado en una caja de zinc que pesó un kilo', 'El pingüino Wenceslao hizo kilómetros bajo exhaustiva lluvia y frío, añoraba a su querido cachorro', 'El veloz murciélago hindú comía feliz cardillo y kiwi. La cigüeña tocaba el saxofón detrás del palenque de paja', 'The quick brown fox jumps over the lazy dog', '', "'Un pangrama es un texto que usa todas las letras posibles del alfabeto de un idioma.'"]


In [26]:
caracteres=0
for i in lista_lineas_pan:
    print(len(i))
    caracteres=caracteres+(len(i))

47
53
120
76
75
98
103
98
111
43
0
86


In [27]:
caracteres

910

In [30]:
# La diferencia entre el número de bytes y el número de caracteres es porque en total hay 11 renglones.
# Saltar de una línea a otra implica el uso de 2 bytes (\n). Esta situación no aplica para los rengones
# vacios.

diferencia=a-caracteres
diferencia

22

## 5.3 Archivos binarios

En cualquier computadora, un archivo es una larga cadena de unos y ceros. En otras palabras, cada archivo es una secuencia finita de bytes. Cada byte es un entero entre 0 y 255 (00000000 y 11111111 en binario). Un archivo binario es diferente a uno de texto porque primero, para poder abrirlo es necesario un programa que sea apropiado al formato, mientras que los archivos de texto se pueden modificar en cualquier editor. En la gran mayoría de los casos, la gente utiliza archivos de tipo binario.

En su escala más pequeña, la información en la computadora es almacenada en bits y bytes. Un **bit** corresponde a la unidad más chica de almacenamiento. Un bit solo puede almacenar un 1 o un 0. Cualquier cosa que tenga 2 estados diferentes se puede almacenar en bits: La carga eléctrica de un chip (0/1), puntos en el disco duro (Norte/Sur). 
Un bit es muy pequeño para poder ser útil por si solo, es por ello que se agrupan 8 de estos para formar **1 byte**.

Un **byte** es una colección de 8 bits. Esto da como resultado expresiones como la siguiente: 01011010. Un byte puede almacenar un caracter; e.g. "A" o "X" o "&". Vamos a analizar como se comportan los bits al agruparlos; consideremos los primeros 3 casos, las combinaciones posibles son las siguientes:

<table>
  <tr>
    <th># de bits</th>
    <th>Combinaciones</th>
    <th>Total</th>
  </tr>
  <tr>
    <td> 1 </td>
    <td> 0  <br><br>  1 </td>
    <td> 2 Combinaciones </td>
  </tr>
  <tr>
    <td> 2 </td>
    <td> 00  <br><br>  01  <br><br>  10  <br><br>  11 </td>
    <td> 4 Combinaciones </td>
  </tr>
  <tr>
    <td> 3 </td>
    <td> 000 . 001  <br><br>  010 . 011  <br><br>  100 . 101  <br><br>  110 . 111 </td>
    <td> 8 Combinaciones </td>
  </tr>
</table>

El número de combinaciones aumenta al doble, por cada bit agregado al conjunto; i.e. 4 bits=16 combinaciones, 5 bits=32 combinaciones,...,etc. Matemáticamente, n bits creará la siguiente cantidad de combinaciones:

$$combinaciones=2^{n bits}$$

Por lo tanto, 1 byte = 8 bits; lo que generará un total de 256 combinaciones. Si a la primera combinación se le asigna una posición de 0, la última combinación tendra una posición de 255. Cada byte servirá para almacenar un caracter diferente. La información generada se almacenará en bytes sin importar el origen o tipo de archivo. Al acumular mil bytes se creará un **Kilobyte (KB)**, un millon de bytes, o lo que es lo mismo, mil KB será un **Megabyte (MB)**, y así sucesivamente.

Los bytes utilizan el código ASCII para representar los diferentes caracteres por un número. Cada número es almacenado en un byte. Por lo tanto, si presionamos **alt+65** obtendremos la letra **"A"**. Eso quiere decir que la letra A está almacenada en el byte 65 en código decimal, o en código binario sería 1000001

La tabla completa de valores **ASCII** se presenta a continuación:

<img src="img\ASCII.png" alt="Python" width="1000"/>

### Actividad

Desarrole una función recursiva para pasar de números decimales a números binarios.

<img src="img/decimal-a-binario.png" alt="Spyder" width="300"/>

<img src="img/spyder.png" alt="Spyder" width="300"/>

In [91]:
def decimal_a_binario(num):
    if num < 2:
        return str(num)
    else:
        return decimal_a_binario(num//2) + str(num % 2)

In [106]:
decimal_a_binario(28)

'11100'

### Clases Byte y Bytearray

Para crear un archivo de bytes desde Python, es necesario utilizar las clases Byte y Bytearray. La primera nos sirve para crear bytes. Un byte servira para almacenar una secuencia entre 0 y 255 (8-bits por su valor en binario).

In [ ]:
# Creamos 4 bytes mediante la instancia de la clase Byte():
cuatro_bytes = bytes(4)

# Revisamos el tipo de objeto creado:
print(type(cuatro_bytes))
print()

# Revisamos el contenido del objeto:
print(cuatro_bytes)

El objeto **cuatro_bytes** equivale a 32 bits de información. La b nos indica que se trata de data de tipo binario. La diagonal invertida junto con la x nos indica que el siguiente valor está en formato hexadecimal. Existen 2 números (00) por byte. Para poder continuar, será necesario utilizar los métodos de conversión entre sistemas decimal-binario-hexadecimal. Para ellos se utilizan los métodos **int(), bin(), hex()** 

In [ ]:
bin(0)

In [ ]:
bin(255)

In [ ]:
bin(100)

In [ ]:
lista_binarios=[]

for i in range(256):
    x=bin(i)
    lista_binarios.append(x)

lista_binarios

In [ ]:
hex(255)

In [ ]:
lista_hexadecimal=[]

for i in range(256):
    x=hex(i)
    lista_hexadecimal.append(x)

lista_hexadecimal

Para poder convertir de binario o hexadecimal a decimal, será necesario usar el método **int()** acompañado de la base del número a convertir:

In [ ]:
binario=bin(10)
print(binario)
print(int(binario,2))

In [ ]:
print(lista_hexadecimal[100])
print(int(lista_hexadecimal[100],16))

Los bytes que sean creados por medio de la clase bytes son **inmutables**. Para poder crear bytes con valor diferente a cero, es necesario seguir la sintaxis que se recibio del contenido de los bytes vacios que se crearon y con valores hexadecimales:

In [ ]:
dos_bytes = bytes(b'\xff\xff')
print(dos_bytes)

Ambos ejemplos, los objetos creados son inmutables. Para crear un objeto mutable, es necesario usar la clase Bytearray:

In [ ]:
# Creamos un arreglo de bytes ya que esto le da la cualidad de ser mutable:
bytes_mutables=bytearray()

# Revisamos el tipo de objeto creado:
print(type(bytes_mutables))
print()

# Revisamos el contenido del objeto:
print(bytes_mutables)

Igual que en el caso anterior, si agregamos un entero a la clase bytearray, nos creará ese número de bytes vacios:

In [ ]:
bytes_mutables=bytearray(3)

print(type(bytes_mutables))
print()

print(bytes_mutables)

In [ ]:
bytes_mutables=bytearray(b'\xAA\xBB\xCC')

print(type(bytes_mutables))
print()

print(bytes_mutables)

Al modificar un arreglo de bytes, es necesario utilizar el valor decimal de nuevo valor que se piensa agregar:

In [ ]:
bytes_mutables[0]=0

In [ ]:
bytes_mutables

In [ ]:
bytes_mutables.append(255)
bytes_mutables

In [ ]:
bytes_mutables[0:2]

In [ ]:
bytes_mutables[3:4]

### Convertir Texto a bytes y viceversa

Para poder convertir texto a bytes y viceversa, se utilizan los métodos **enconde()** y **decode()**.

In [116]:
texto="Edgar"
t=texto.encode()

In [126]:
print(type(texto))
print(type(t))

<class 'str'>
<class 'bytes'>


In [127]:
print(texto)
print(t)

Edgar
b'Edgar'


el método **encode()** permite transformar el contenido de un objeto string de una codificación en otra. El resultado de la transformación es un objetos tipo bytes. El valor por defecto de encoding es **"utf-8"**. Existen diferentes tipos de encodings para utilizar. Un ejemplo es el código **ascii**. Si se trata de codificar un texto, que utilice caracteres que no esten el la lista ascii, generará un error como se muestra a continuación:

In [125]:
#nuevo_texto="Niño".encode(encoding="ascii")
#nuevo_texto

b'Ni?o'

In [130]:
# Para manejar ese y otros errores al usar encoding, podemos usar el kwarg errors:
nuevo_texto="Niño".encode(encoding="ascii", errors="ignore")
nuevo_texto

b'Nio'

Existen diferentes opciones para el **kwarg errors**. Las más comunes son:
* strict: Forma por defento de encode, generará error si no puede manejar el encoding
* ignore: Ignora el caracter en cuestión y prosigue al siguiente
* replace: Sustituye al caracter por un signo de interrogación
* xmlcharrefreplace: Sustituye al caracter por el valor correspondiente XML

In [131]:
nuevo_texto="Niño".encode(encoding="ascii", errors="replace")
nuevo_texto

b'Ni?o'

Si mandamos a llamar byte por byte, obtendremos su valor en código ascii:

In [133]:
nuevo_texto[0]

78

In [134]:
nuevo_texto[3]

111

In [135]:
nuevo_texto[2]

63

In [136]:
#El método decode funciona exactamente igual que el método encode. En este caso damos el valor utf-8
string=nuevo_texto.decode(encoding="utf-8")
string

'Ni?o'

In [140]:
# Tambien es posible convertir un objeto de bytes a formato hexadecimal:
nuevo_texto.hex()

'4e693f6f'

In [143]:
# La sintaxis inversa se presenta a continuación:
b"".fromhex("4e693f6f")

b'Ni?o'